### DISCUSSION TITLE
Mean implementation for Fixed-Points

### DISCUSSION THREAD

**Discussion by manel1874** [lschoe/mpyc#74] [Q&A] [answered]
Hi, I am curious why we have the following implementation for the mean over FP:

https://github.com/lschoe/mpyc/blob/b8d2d5d6d681560bba062f7b2f54fa0734e90d29/mpyc/statistics.py#L59

Is that because it's faster to divide a secret number by a power of two rather than by some general number?

**Comment by lschoe** [comment, marked as answer]
It's to prevent (unnecessary) loss of precision in the division by the public integer `n`. If we simply divide by `n`, which works out as multiplying by `1/n`, we get a fixed-point number of the form `0.00001***` say, in binary, with a lot of leading zeros and few significant bits. Therefore, we shift this to the left to make it of the form `0.1*******`, which represents `1/n` with more significant bits. The product with the sum `s` is shifted back at the end. Note that `e = n.bit_length()-1` to ensure `1/2 < 2**e / n <= 1` as written in the comment on line 58.

**Reply by manel1874** [reply]
Great, thanks for you fast reply!
